csrc/multidevice/executor.cpp

```diff
@@ -49,46 +51,95 @@ std::unique_ptr<Fusion> copyFusionAndChangeOutputs(
 MultiDeviceExecutor::MultiDeviceExecutor(
     std::unique_ptr<Fusion> fusion,
     Communicator& comm,
-    MultiDeviceExecutorParams params)
-    : comm_(comm), params_(params) {
+    hir::HostIrExecutorParams params)
+    : comm_(comm), complete_fusion_(std::move(fusion)) {
   // Sharding PreSegmenter passes.
   // Note: passes run before PreSegmenter optimization passes.
-  propagateShardings(fusion.get());
-  insertReshardings(fusion.get());
-  insertShardedAxisReordering(fusion.get());
-  setShardedAllocationDomain(fusion.get());
+  propagateShardings(complete_fusion_.get());
+  insertReshardings(complete_fusion_.get());
+  insertShardedAxisReordering(complete_fusion_.get());
+  setShardedAllocationDomain(complete_fusion_.get());
+
+  // Performs segmentation at the inter-device communications
+  // Each SegmentedGroup represents a pipeline's stage, and can be either
+  // 1) a Fusion which doesn't involve inter-device communication
+  // 2) a Fusion comprised of one Expr, representing inter-device communication
   SegmentCandidateFinderOptions options{
       .run_translate_welford = false,
       .run_combine_reductions = false,
       .run_herrmann_merge = true,
       .run_final_merge = true,
       .only_segment_resharding_exprs = true};
+  std::unique_ptr<SegmentedFusion> staged_fusion =
+      SegmentCandidateFinder::segment(
+          std::make_unique<Fusion>(*complete_fusion_), nullptr, options);
+  // Infer a topologically ordered traversal of the segmented fusion to
+  // determine the order for launching the kernels/comms
+  RuntimeWorkSpace workspace;
+  prepareRuntimeOrder(staged_fusion.get(), workspace);
+
+  // Create the HostIrContainer representing the host program. Each segment of
+  // the segmented fusion will be translated to a HostIR
+  auto hic = std::make_unique<hir::HostIrContainer>();
+  FusionGuard fg(hic.get());
+  IrCloner ir_cloner(hic.get());
+  auto clone =
+      [&ir_cloner](const std::vector<Val*>& vals) -> std::vector<Val*> {
+    std::vector<Val*> cloned_vals(vals.size());
+    std::transform(
+        vals.begin(), vals.end(), cloned_vals.begin(), [&ir_cloner](Val* val) {
+          return ir_cloner.clone(val);
+        });
+    return cloned_vals;
+  };
 
-  staged_fusion_ =
-      SegmentCandidateFinder::segment(std::move(fusion), nullptr, options);
-
-  for (auto group : staged_fusion_->groups()) {
+  for (auto group : workspace.group_run_order) {
+    std::vector<Expr*> host_exprs;
     NVF_ERROR(!group->exprs().empty(), "invalid segmentation");
-    is_resharding_[group] = std::any_of(
+    bool is_resharding = std::any_of(
```

Comment:
```suggestion
    const bool is_resharding = std::any_of(
```

Reply:
Apparently, this can be moved after the involvedDevices check so `is_resharding` gets closer to where it's used for readability.

Reply:
done!